### ISSUE TITLE
Failed in nopython mode pipeline len(none)

### ISSUE THREAD

**Issue by nehaljwani** [numba/numba#7482]
I'm on the latest build of numba:
```bash
$ conda list numba
# packages in environment at /path/to/my/env:
#
# Name                    Version                   Build  Channel
numba                     0.54.1           py37h2d894fd_0    conda-forge
```

Small reproducible example:
```python
$ cat bug.py 
import math
import numba

@numba.jit(nopython=True, fastmath=True)
def compute(smth, weights, default=0.0):

    if weights is None:
        return None

    if len(weights) == 0:
        return default

    idx = smth > 0
    weights = weights[idx]

    return default*weights

compute(smth=1, weights=None)
```

The problem:
```bash
$ python bug.py 
Traceback (most recent call last):
  File "bug.py", line 19, in <module>
    compute(smth=1, weights=None)
  File "/path/to/my/env/lib/python3.7/site-packages/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/path/to/my/env/lib/python3.7/site-packages/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function len>) found for signature:
 
 >>> len(none)
 
There are 16 candidate implementations:
  - Of which 16 did not match due to:
  Overload of function 'len': File: <numerous>: Line N/A.
    With argument(s): '(none)':
   No match.

During: resolving callee type: Function(<built-in function len>)
During: typing of call at bug.py (10)


File "bug.py", line 10:
def compute(smth, weights, default=0.0):
    <source elided>

    if len(weights) == 0:
    ^
```

This diff resolves the problem, but I didn't need it with `numba=0.43.1`
```diff
--- bug.py	2021-10-13 16:14:49.090431600 -0400
+++ /tmp/bug.py	2021-10-13 16:17:51.534177904 -0400
@@ -11,9 +11,9 @@
         return default
 
     idx = smth > 0
-    weights = weights[idx]
+    weight = weights[idx]
 
-    return default*weights
+    return default*weight
 
 
 compute(smth=1, weights=None)
 ```

**Comment by esc** [comment]
@nehaljwani thank you for reporting this, this will need looking at.

**Comment by esc** [comment]
> Git bisect reveals that the following commit is the inflection point:
> 
> ```
> commit 24fcb48f683f16504ab26fef08093f92a43ba7e6 (refs/bisect/bad)
> Author: Stuart Archibald <stuart.archibald@googlemail.com>
> Date:   Tue Jun 11 16:35:36 2019 +0100
> 
>     Abandon branch pruning if an arg name is redefined.
> 
>     This makes it so that branch pruning for a given input arg is
>     abandoned if an input arg is redefined in the user code as its
>     value potentially becomes non-const (and no const-prop analysis
>     yet to figure it out).
> 
>     Fixes #4163
> ```

@nehaljwani thank you very much for your efforts to bisect this. The question now is, @stuartarchibald is this a genuine bug, or is it something that "used to work by accident"?

**Comment by stuartarchibald** [comment]
@nehaljwani / @esc Note: I redacted some data in the above but otherwise is the same.

@nehaljwani thanks for bisecting. This is a mixture of adding a feature to prune dead branches and lack of SSA at that point in the pipeline, which means that if an arg is redefined the pruner has to abort as it cannot work out whether it is rewritten or not as the name is the same. The patch 24fcb48 fixed https://github.com/numba/numba/issues/4163 which has some more details. I'm afraid this cannot be fixed until the pipeline is SSA compliant from the start, this is something that's being worked on (though progress is somewhat slow as it's challenging)!
